Introduce test suite

mikapfl opened the pull request:
# Purpose of this PR

So far, we tell people to run specific tests manually in each pull request (`Rscript start.R -g config/scenario_config_AMT.csv` and `Rscript start.R -q`), and additionally we have some ideas / information how to run other types of tests. I propose in this PR to introduce a test suite as a place to collect tests so that it is easier to run the tests and failure and success are easily visible.

I propose to categorize the tests into "fast" tests and "slow" tests, where the goal is that "fast" tests finish within about 10 minutes on a laptop, and the slow tests take as long as it takes. The fast tests can be run using `make test`, the slow tests using `make test-slow`.

For starters, I translated `start.R -q` and `start.R -g config/scenario_config_AMT.csv` into tests (`test_1-quick.R` and `test_2-gamscompile.R`, respectively) for the default testing suite and a `start.R` with the default config as `test_3-default.R` for the slow testing suite.

Additional tests (in particular, unit tests for functions and scripts, or maybe even standalone module tests, or integration tests for REMIND-MAgPIE coupling) can be added in the future.

## Type of change

- [x] New feature 
- [x] This change requires a documentation update


## Checklist:

- [x] My code follows the coding etiquette
- [x] I have performed a self-review of my own code
- [x] Changes are commented, particularly in hard-to-understand areas
- [x] I have updated the in-code documentation
- [x] I have adjusted reporting where it was needed
- [x] Tests are passing (`make test` shows no FAILed tests)
- [x] documentation (tutorials) and github PR templates are updated

mikapfl commented:
This is the results of the tests if you forget a semi-colon in main.gms:

```
══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 414.8 s

── Skipped tests  ───────────────────────────────────────────────────────────────────────────────────────────────────────────
• Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 2 | WARN 0 | SKIP 1 | PASS 0 ]
Error: Test failures
Execution halted
make: *** [Makefile:32: test] Error 1
```

This is the result if tests pass (in the "fast" testing suite, with the "slow" tests skipped):
```
══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 596.9 s

── Skipped tests  ───────────────────────────────────────────────────────────────────────────────────────────────────────────
• Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 0 | WARN 0 | SKIP 1 | PASS 2 ]
```

orichters commented on config/tests/scenario_config_default.csv:
```diff
@@ -0,0 +1,2 @@
+title;start;slurmConfig
+default;1;--qos=standby --nodes=1 --tasks-per-node=12 --wait
```
if I understand correctly, [based on this document](https://slurm.schedmd.com/slurm.conf.html#OPT_WaitTime) `--wait`, it means that `start.R` waits until the run is ready? If I understand it correctly, this option is limited to 18 hours, 20 minutes anyway, so you may also set `--time 1092` which may increase the chances of the run starting early.

mikapfl replied:
slurm is confusing, this is not `srun --wait=x`, this is [`sbatch --wait`](https://slurm.schedmd.com/sbatch.html#OPT_wait), which means it just waits for the job to be finished, then exits. So start.R waits until the run is finished, whatever it takes. It is needed because otherwise start.R just submits the job and then exits successfully.

orichters commented on scripts/start/submit.R:
```diff
@@ -106,13 +108,15 @@ submit <- function(cfg, restart = FALSE, stopOnFolderCreateError = TRUE) {
+    exitCode <- system("Rscript prepare_and_run.R")
   } else {
-    system(paste0("sbatch --job-name=",cfg$title," --output=log.txt --mail-type=END --comment=REMIND --wrap=\"Rscript prepare_and_run.R \" ",cfg$slurmConfig))
+    exitCode <- system(paste0("sbatch --job-name=",cfg$title," --output=log.txt --mail-type=END --comment=REMIND --wrap=\"Rscript prepare_and_run.R \" ",cfg$slurmConfig))
```
So this means in case we have multiple scenarios in the test/config files, they would not run in parallel, but one after the other?

mikapfl replied:
If we specify multiple scenarios in test config files and they specify `--wait` in the slurm config, then they are not run in parallel, that's correct. When we have more tests and want to run them in parallel, we'll need to figure something out.

mikapfl commented on start.R:
```diff
@@ -193,14 +195,19 @@ if ("--gamscompile" %in% flags) {
-      if (gms::getLine() %in% c("Y", "y")) {
-        runGamsCompile(modelFile, cfg, interactive)
+    if (0 < exitcode) {
+      ignorederrors <<- ignorederrors + 1
```
I'm not happy with it either.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q replied:
Why?  It's perfectly fine.
Alternatively, return the number of ignored errors and increase outside the function.  Much more hassle.

orichters replied:
If I remember correctly, the argument was that it makes debugging more difficult and may create problems if at some point the function is reused in some other code or put into a different file or so.

mikapfl replied:
The problem is generally that it breaks the mental model of "you only have to look at the function's definition to know how it is dealing with the rest of the code". That's why - like goto in other programming languages - usage of global variables and other global state is often a code smell (meaning, it is not inherently wrong, but it is more often than not an indication that something is wrong). Here, I would argue that the job of several functions in the start scripts is not clearly defined and that is the reason why we resort to global variables. Initially, I introduced another variable (named "failed") and only later found that we already *have* a global variable for this purpose. That is one example why global variables are difficult to maintain - they are difficult to even know about.

But anyhow, using this global variable which already exists in another place doesn't make things substantially worse, and I don't feel comfortable refactoring the start scripts at the moment without reliable tests in place.

orichters commented on config/tests/scenario_config_quick.csv:
```diff
@@ -0,0 +1,2 @@
+title;start;cm_nash_mode;cm_iteration_max;optimization;results_folder;output;cm_quick_mode;force_replace;slurmConfig
+testOneRegi;1;debug;1;testOneRegi;output/testOneRegi;NA;on;TRUE;--qos=priority --nodes=1 --tasks-per-node=1 --mem=8000 --time=60 --wait
```
yeah, but
```
> isTRUE("TRUE")
[1] FALSE
```

mikapfl commented:
> * `gms::codeCheck(strict = TRUE)` might be another thing to add?

Hm, we already have it in `make check` and people can choose between `make check-fix` for interactive use and the non-interactive variant. The non-interactive varaint takes about 1 minute, so I guess we could put it into `make test` as well, just to be sure people get it if the only thing they run is `make test`.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
Doesn't work locally.
```
$ make test
R_PROFILE_USER= Rscript -e 'testthat::test_dir("tests/testthat")'
✔ | F W S  OK | Context
⠏ |         0 | 0-codeCheck                                                                
 Running codeCheck...
 Finished data collection...            (time elapsed:   3.03)
 Naming conventions check done...       (time elapsed:   3.03)
  Running checkAppearance...
  Start variable matching...            (time elapsed:  0.073)
  Finished variable matching...         (time elapsed:  69.93)
 Investigated variable appearances...   (time elapsed:  73.04)
 Appearance and usage check done...     (time elapsed:  73.10)
 Switch Appearance check done...        (time elapsed:  73.51)
 Interface collection and check done... (time elapsed:  73.59)
 Input folder check done...             (time elapsed:  73.59)
 Description check done...              (time elapsed:  73.59)
 All codeCheck tests passed!
✔ |         1 | 0-codeCheck [81.3s]                                                        
⠏ |         0 | 1-quick                                                                    The following updates are available:
- edgeTransport : 0.22.3 -> 0.22.5
- lucode2 : 0.37.1 -> 0.39.0
- magpie4 : 1.165.7 -> 1.166.0
- mip : 0.138.1 -> 0.139.1
- mrcommons : 1.23.3 -> 1.23.11
- mrland : 0.40.0 -> 0.40.3
- mrremind : 0.148.0.9001 -> 0.148.1
- mrvalidation : 2.44.1 -> 2.45.3
- mrwater : 1.3.35 -> 1.4.14
- piamInterfaces : 0.0.26 -> 0.0.31
- quitte : 0.3103.0 -> 0.3105.0
- remind2 : 1.103.13 -> 1.103.16
Consider updating with `Rscript scripts/utils/updateRenv.R`.

Reading config file config/tests/scenario_config_quick.csv 

testOneRegi
   Configuring cfg for testOneRegi
   Run can be started using 0 specified gdx file(s).
   Writing cfg to file testOneRegi.RData
    Results folder already exists, deleting and re-creating it: output/testOneRegi

The following updates are available:
- edgeTransport : 0.22.3 -> 0.22.5
- lucode2 : 0.37.1 -> 0.39.0
- magpie4 : 1.165.7 -> 1.166.0
- mip : 0.138.1 -> 0.139.1
- mrcommons : 1.23.3 -> 1.23.11
- mrland : 0.40.0 -> 0.40.3
- mrremind : 0.148.0.9001 -> 0.148.1
- mrvalidation : 2.44.1 -> 2.45.3
- mrwater : 1.3.35 -> 1.4.14
- piamInterfaces : 0.0.26 -> 0.0.31
- quitte : 0.3103.0 -> 0.3105.0
- remind2 : 1.103.13 -> 1.103.16
Consider updating with `Rscript scripts/utils/updateRenv.R`.
   Initializing renv, see output/testOneRegi/log_renv.txt
   Writing cfg to file output/testOneRegi/config.Rdata 
   Executing prepare_and_run.R for output/testOneRegi 
Loading required package: gdx
Loading required package: gdxrrw
Loading required package: magclass
Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 
Loading required package: mrremind
Loading required package: madrat
Loading required package: mrcommons
Loading required package: mrdrivers
Loading required package: mstools
Loading required package: edgeTransport
Loading required package: data.table

Attaching package: ‘data.table’

The following objects are masked from ‘package:dplyr’:

    between, first, last

Loading required package: mrvalidation
Loading required package: mrmagpie
Loading required package: mrland
Loading required package: mrwater

===== git info =====
Latest commit: 6d806c3 2022-11-29 16:54:58 +0100 Mika Pflüger 
Changes since then: On branch mika-test
Your branch is up to date with 'mika-github/test'.

nothing to commit, working tree clean

==== installed package versions =====
# A tibble: 27 × 2
   Package       Version     
   <chr>         <chr>       
 1 data.table    1.14.4      
 2 devtools      2.4.5       
 3 dplyr         1.0.10      
 4 edgeTransport 0.22.3      
 5 flexdashboard 0.6.0       
 6 gdx           1.51.5      
 7 gdxdt         0.1.0       
 8 gdxrrw        1.0.10      
 9 ggplot2       3.3.6       
10 gtools        3.9.3       
11 lucode2       0.37.1      
12 luplot        3.54.2      
13 luscale       2.26.5      
14 magclass      6.6.0       
15 magpie4       1.165.7     
16 mip           0.138.1     
17 mrremind      0.148.0.9001
18 mrvalidation  2.44.1      
19 optparse      1.7.3       
20 plotly        4.10.0      
21 remind2       1.103.13    
22 rlang         1.0.6       
23 rmndt         0.5.4       
24 tidyverse     1.3.2       
25 methods       4.2.1       
26 parallel      4.2.1       
27 tools         4.2.1       

==========
Copying MAGICC files from /home/pehl/PIK/magicc/ to results folder
2022-11-29 17:06:00: acquiring model lock...
acquired model lock in 0 secs.
Your input data are outdated or in a different regional resolution. New input data are downloaded and distributed.
Delete old data in input folders ... 
done!

Load data..
  try ~/PIK/Remind_input
Error in download_unpack(input = files, targetdir = "input", repositories = repositories,  : 
  Following files not found:
  rev6.324_62eff8f7_remind.tgz
  rev6.324_62eff8f7_validationremind.tgz
  CESparametersAndGDX_34a93aa9d54616d1c3b302aa3002ec161da24a94.tgz
Calls: prepare -> download_distribute -> download_unpack
2022-11-29 17:06:01: unlocked model.
Execution halted
Error in submit(cfg) : Executing prepare_and_run failed, stopping.
Execution halted
✖ | 1       0 | 1-quick [44.8s]                                                            
───────────────────────────────────────────────────────────────────────────────────────────
Failure (test_1-quick.R:3): start.R config/tests/scenario_config_quick.csv works
`status` not equal to 0.
1/1 mismatches
[1] 1 - 0 == 1
───────────────────────────────────────────────────────────────────────────────────────────
⠏ |         0 | 2-gamscompile                                                              The following updates are available:
- edgeTransport : 0.22.3 -> 0.22.5
- lucode2 : 0.37.1 -> 0.39.0
- magpie4 : 1.165.7 -> 1.166.0
- mip : 0.138.1 -> 0.139.1
- mrcommons : 1.23.3 -> 1.23.11
- mrland : 0.40.0 -> 0.40.3
- mrremind : 0.148.0.9001 -> 0.148.1
- mrvalidation : 2.44.1 -> 2.45.3
- mrwater : 1.3.35 -> 1.4.14
- piamInterfaces : 0.0.26 -> 0.0.31
- quitte : 0.3103.0 -> 0.3105.0
- remind2 : 1.103.13 -> 1.103.16
Consider updating with `Rscript scripts/utils/updateRenv.R`.

Reading config file config/scenario_config_AMT.csv 

No column path_gdx_refpolicycost for policy cost comparison found, using path_gdx_ref instead.

Trying to compile the selected runs...
2022-11-29 17:06:08: acquiring model lock...
acquired model lock in 0 secs.
FAIL output/gamscompile/main_SSP2EU-AMT-calibrate.lst
FAIL output/gamscompile/main_testOneRegi-AMT-Base.lst
FAIL output/gamscompile/main_SSP1-AMT-Base.lst
FAIL output/gamscompile/main_SSP1-AMT-NDC.lst
FAIL output/gamscompile/main_SSP1-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SSP1-AMT-PkBudg1320.lst
FAIL output/gamscompile/main_SSP2EU-AMT-Base.lst
FAIL output/gamscompile/main_SSP2EU-AMT-NDC.lst
FAIL output/gamscompile/main_SSP2EU-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SSP2EU-AMT-PkBudg1320.lst
FAIL output/gamscompile/main_SSP5-AMT-Base.lst
FAIL output/gamscompile/main_SSP5-AMT-NDC.lst
FAIL output/gamscompile/main_SSP5-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SSP5-AMT-PkBudg1320.lst
FAIL output/gamscompile/main_SDP-AMT-Base.lst
FAIL output/gamscompile/main_SDP-AMT-NDC.lst
FAIL output/gamscompile/main_SDP-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SDP-AMT-PkBudg1320.lst

2022-11-29 17:06:35: unlocked model.

Finished: 18 runs started. 0 runs are waiting. 
To investigate potential FAILs, run: less -j 4 --pattern='^\*\*\*\*' filename.lst
Error: 18 runs were not started successfully.
Execution halted
✖ | 1       0 | 2-gamscompile [33.2s]                                                      
───────────────────────────────────────────────────────────────────────────────────────────
Failure (test_2-gamscompile.R:3): start.R --gamscompile config/scenario_config_AMT.csv works
`status` not equal to 0.
1/1 mismatches
[1] 1 - 0 == 1
───────────────────────────────────────────────────────────────────────────────────────────
✔ |     1   0 | 3-default                                                                  
───────────────────────────────────────────────────────────────────────────────────────────
Skip (test_3-default.R:2): start.R works
Reason: Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes).
───────────────────────────────────────────────────────────────────────────────────────────

══ Results ════════════════════════════════════════════════════════════════════════════════
Duration: 159.4 s

── Skipped tests  ─────────────────────────────────────────────────────────────────────────
• Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 2 | WARN 0 | SKIP 1 | PASS 1 ]

Keep trying!
Error: Test failures
Execution halted
make: *** [Makefile:32: test] Error 1
```

mikapfl commented:
Doesn't work locally because you haven't set up your local REMIND to download input data, see https://github.com/remindmodel/remind/blob/develop/tutorials/02a_RunningREMINDLocally.md .

mikapfl commented:
With input data set up, it works:

```
pflueger@lt487 > make test
R_PROFILE_USER= Rscript -e 'testthat::test_dir("tests/testthat")'
✔ | F W S  OK | Context
⠏ |         0 | 0-codeCheck                                                                                                  
 Running codeCheck...
 Finished data collection...            (time elapsed:   1.24)
 Naming conventions check done...       (time elapsed:   1.24)
  Running checkAppearance...
  Start variable matching...            (time elapsed:  0.044)
  Finished variable matching...         (time elapsed:  53.50)
 Investigated variable appearances...   (time elapsed:  54.77)
 Appearance and usage check done...     (time elapsed:  54.83)
 Switch Appearance check done...        (time elapsed:  55.17)
 Interface collection and check done... (time elapsed:  55.23)
 Input folder check done...             (time elapsed:  55.23)
 Description check done...              (time elapsed:  55.23)
 All codeCheck tests passed!
✔ |         1 | 0-codeCheck [56.3s]                                                                                          
⠏ |         0 | 1-quick                                                                                                      
Reading config file config/tests/scenario_config_quick.csv 

testOneRegi
   Configuring cfg for testOneRegi
   Run can be started using 0 specified gdx file(s).
   Writing cfg to file testOneRegi.RData
    Results folder already exists, deleting and re-creating it: output/testOneRegi

   Initializing renv, see output/testOneRegi/log_renv.txt
   Writing cfg to file output/testOneRegi/config.Rdata 
   Executing prepare_and_run.R for output/testOneRegi 
Loading required package: gdx
Loading required package: gdxrrw
Loading required package: magclass
Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 
Loading required package: mrremind
Loading required package: madrat
Loading required package: mrcommons
Loading required package: mrdrivers
Loading required package: mstools
Loading required package: edgeTransport
Loading required package: data.table

Attaching package: ‘data.table’

The following objects are masked from ‘package:dplyr’:

    between, first, last

Loading required package: mrvalidation
Loading required package: mrmagpie
Loading required package: mrland
Loading required package: mrwater

===== git info =====
Latest commit: 4f969a1c 2022-11-29 16:57:10 +0100 Mika Pflüger 
Changes since then: On branch faster-compile-test
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	../../todo

nothing added to commit but untracked files present (use "git add" to track)

==== installed package versions =====
# A tibble: 27 × 2
   Package       Version 
   <chr>         <chr>   
 1 data.table    1.14.6  
 2 devtools      2.4.5   
 3 dplyr         1.0.10  
 4 edgeTransport 0.22.5  
 5 flexdashboard 0.6.0   
 6 gdx           1.51.5  
 7 gdxdt         0.1.0   
 8 gdxrrw        1.0.10  
 9 ggplot2       3.4.0   
10 gtools        3.9.3   
11 lucode2       0.39.0  
12 luplot        3.54.2  
13 luscale       2.26.5  
14 magclass      6.6.0   
15 magpie4       1.166.0 
16 mip           0.139.1 
17 mrremind      0.148.1 
18 mrvalidation  2.45.3  
19 optparse      1.7.3   
20 plotly        4.10.1  
21 remind2       1.103.16
22 rlang         1.0.6   
23 rmndt         0.5.4   
24 tidyverse     1.3.2   
25 methods       4.2.2   
26 parallel      4.2.2   
27 tools         4.2.2   

==========
Copying MAGICC files from /p/projects/rd3mod/magicc/ to results folder
2022-11-29 16:58:44: acquiring model lock...
acquired model lock in 0 secs.
No input data downloaded and distributed. To enable that, delete input/source_files.log or set cfg$force_download to TRUE.

Creating full.gms
2022-11-29 16:58:59: unlocked model.
Saving timePrepareStart and timePrepareEnd to runstatistics.rda

Starting REMIND...
GAMS will provide logging in full.log.
Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 
[1] "--- 2022-11-29 17:01:53 Start of the EDGE-T iterative model run."
[1] "Loading csv data from input folder and creating RDS files..."
[1] "Fuel prices lower than 1$/GJ found. Regions: EUR, MEA"
[1] "The weighted averages of non-zero regions of the corresponding fuel will be used."
[1] "Iterative logit calculation finished in 0.128375152746836 Minutes"
NULL

REMIND run finished!

Model summary:
  gams_runtime is 3.4 mins.
  full.gms exists, so the REMIND GAMS code was generated.
  full.log and full.lst exist, so GAMS did run.
  abort.gdx does not exist, a file written automatically for some types of errors.
  non_optimal.gdx does not exist, a file written if at least one iteration did not find a locally optimal solution.
  fulldata.gdx exists, because iteration 1 was successful. modelstat: 2 (Locally Optimal)
  Modelstat after 1 iterations: 2 (Locally Optimal)
  full.log states: *** Status: Normal completion


Collect and submit run statistics to central data base.
Cannot access run statistics repository /p/projects/rd3mod/models/statistics/remind Run statistics are not submitted.

Did not try to start subsequent runs.

Loading required package: stringr

Starting output generation for output/testOneRegi


Finished output generation for output/testOneRegi!


Saving timeGAMSStart, timeGAMSEnd, timeOutputStart and timeOutputStart to runstatistics.rda
[1] "output/testOneRegi"


Finished: 1 runs started. 0 runs are waiting. 
✔ |         1 | 1-quick [243.9s]                                                                                             
⠏ |         0 | 2-gamscompile                                                                                                
Reading config file config/scenario_config_AMT.csv 

No column path_gdx_refpolicycost for policy cost comparison found, using path_gdx_ref instead.

Trying to compile the selected runs...
2022-11-29 17:02:26: acquiring model lock...
acquired model lock in 0 secs.
  OK output/gamscompile/main_SSP2EU-AMT-calibrate.lst
  OK output/gamscompile/main_testOneRegi-AMT-Base.lst
  OK output/gamscompile/main_SSP1-AMT-Base.lst
  OK output/gamscompile/main_SSP1-AMT-NDC.lst
  OK output/gamscompile/main_SSP1-AMT-PkBudg670.lst
  OK output/gamscompile/main_SSP1-AMT-PkBudg1320.lst
  OK output/gamscompile/main_SSP2EU-AMT-Base.lst
  OK output/gamscompile/main_SSP2EU-AMT-NDC.lst
  OK output/gamscompile/main_SSP2EU-AMT-PkBudg670.lst
  OK output/gamscompile/main_SSP2EU-AMT-PkBudg1320.lst
  OK output/gamscompile/main_SSP5-AMT-Base.lst
  OK output/gamscompile/main_SSP5-AMT-NDC.lst
  OK output/gamscompile/main_SSP5-AMT-PkBudg670.lst
  OK output/gamscompile/main_SSP5-AMT-PkBudg1320.lst
  OK output/gamscompile/main_SDP-AMT-Base.lst
  OK output/gamscompile/main_SDP-AMT-NDC.lst
  OK output/gamscompile/main_SDP-AMT-PkBudg670.lst
  OK output/gamscompile/main_SDP-AMT-PkBudg1320.lst

2022-11-29 17:05:57: unlocked model.

Finished: 18 runs started. 0 runs are waiting. 
To investigate potential FAILs, run: less -j 4 --pattern='^\*\*\*\*' filename.lst
✔ |         1 | 2-gamscompile [215.4s]                                                                                       
✔ |     1   0 | 3-default                                                                                                    
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Skip (test_3-default.R:2): start.R works
Reason: Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes).
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 515.7 s

── Skipped tests  ───────────────────────────────────────────────────────────────────────────────────────────────────────────
• Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 0 | WARN 0 | SKIP 1 | PASS 3 ]
```

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> Doesn't work locally because you haven't set up your local REMIND to download input data, see https://github.com/remindmodel/remind/blob/develop/tutorials/02a_RunningREMINDLocally.md .

Of course I did.  I wrote that code.
I just missed out on the data update and didn't download yet.

mikapfl commented:
> I just missed out on the data update and didn't download yet.

Huh, also in the tests, that should work automatically, I think. Or do you mean, you are using a local repo, and it doesn't contain the latest data update? Because if you use scp to the cluster, I think it would be a bug in the tests if they don't download that input data.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> > I just missed out on the data update and didn't download yet.
> 
> Huh, also in the tests, that should work automatically, I think. Or do you mean, you are using a local repo, and it doesn't contain the latest data update? Because if you use scp to the cluster, I think it would be a bug in the tests if they don't download that input data.

Call it manual caching.  During development/debugging, I switch input data revisions often, and I don't want to download anew them every time.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> This is the results of the tests if you forget a semi-colon in main.gms:
> 
> ```
> ══ Results ══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
> Duration: 414.8 s
> 
> ── Skipped tests  ───────────────────────────────────────────────────────────────────────────────────────────────────────────
> • Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes). (1)
> 
> [ FAIL 2 | WARN 0 | SKIP 1 | PASS 0 ]
> Error: Test failures
> Execution halted
> make: *** [Makefile:32: test] Error 1
> ```

No.
THIS is the output if you forget a semi-colon in `main.gms`:
```
$ make test
R_PROFILE_USER= Rscript -e 'testthat::test_dir("tests/testthat")'
✔ | F W S  OK | Context
⠏ |         0 | 0-codeCheck                                                                
 Running codeCheck...
 Finished data collection...            (time elapsed:   1.76)
 Naming conventions check done...       (time elapsed:   1.77)
  Running checkAppearance...
  Start variable matching...            (time elapsed:  0.069)
  Finished variable matching...         (time elapsed:  68.24)
 Investigated variable appearances...   (time elapsed:  70.08)
 Appearance and usage check done...     (time elapsed:  70.14)
 Switch Appearance check done...        (time elapsed:  70.55)
 Interface collection and check done... (time elapsed:  70.62)
 Input folder check done...             (time elapsed:  70.62)
 Description check done...              (time elapsed:  70.62)
 All codeCheck tests passed!
✔ |         1 | 0-codeCheck [72.1s]                                                        
⠏ |         0 | 1-quick
Reading config file config/tests/scenario_config_quick.csv 

testOneRegi
   Configuring cfg for testOneRegi
   Run can be started using 0 specified gdx file(s).
   Writing cfg to file testOneRegi.RData
    Results folder already exists, deleting and re-creating it: output/testOneRegi

   Initializing renv, see output/testOneRegi/log_renv.txt
   Writing cfg to file output/testOneRegi/config.Rdata 
   Executing prepare_and_run.R for output/testOneRegi 
Loading required package: gdx
Loading required package: gdxrrw
Loading required package: magclass
Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 
Loading required package: mrremind
Loading required package: madrat
Loading required package: mrcommons
Loading required package: mrdrivers
Loading required package: mstools
Loading required package: edgeTransport
Loading required package: data.table

Attaching package: ‘data.table’

The following objects are masked from ‘package:dplyr’:

    between, first, last

Loading required package: mrvalidation
Loading required package: mrmagpie
Loading required package: mrland
Loading required package: mrwater

===== git info =====
Latest commit: 6d806c3 2022-11-29 16:54:58 +0100 Mika Pflüger 
Changes since then: On branch mika-test
Your branch is up to date with 'mika-github/test'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ../../main.gms

no changes added to commit (use "git add" and/or "git commit -a")

==== installed package versions =====
# A tibble: 27 × 2
   Package       Version     
   <chr>         <chr>       
 1 data.table    1.14.4      
 2 devtools      2.4.5       
 3 dplyr         1.0.10      
 4 edgeTransport 0.22.3      
 5 flexdashboard 0.6.0       
 6 gdx           1.51.5      
 7 gdxdt         0.1.0       
 8 gdxrrw        1.0.10      
 9 ggplot2       3.3.6       
10 gtools        3.9.3       
11 lucode2       0.37.1      
12 luplot        3.54.2      
13 luscale       2.26.5      
14 magclass      6.6.0       
15 magpie4       1.165.7     
16 mip           0.138.1     
17 mrremind      0.148.0.9001
18 mrvalidation  2.44.1      
19 optparse      1.7.3       
20 plotly        4.10.0      
21 remind2       1.103.13    
22 rlang         1.0.6       
23 rmndt         0.5.4       
24 tidyverse     1.3.2       
25 methods       4.2.1       
26 parallel      4.2.1       
27 tools         4.2.1       

==========
Copying MAGICC files from /home/pehl/PIK/magicc/ to results folder
2022-11-30 10:43:36: acquiring model lock...
acquired model lock in 0 secs.
No input data downloaded and distributed. To enable that, delete input/source_files.log or set cfg$force_download to TRUE.

Creating full.gms
2022-11-30 10:46:39: unlocked model.
Saving timePrepareStart and timePrepareEnd to runstatistics.rda

Starting REMIND...
GAMS will provide logging in full.log.

REMIND run finished!

Model summary:
  gams_runtime is 12.7 secs.
  full.gms exists, so the REMIND GAMS code was generated.
  full.log and full.lst exist, so GAMS did run.
  abort.gdx does not exist, a file written automatically for some types of errors.
  non_optimal.gdx does not exist, a file written if at least one iteration did not find a locally optimal solution.
! fulldata.gdx does not exist, so output generation will fail.
  full.log states: *** Status: Compilation error(s)


Collect and submit run statistics to central data base.
Cannot access run statistics repository /p/projects/rd3mod/models/statistics/remind Run statistics are not submitted.
Error in run(start_subsequent_runs) : 
  GAMS did not complete its run, so stopping here:
       No output is generated, no subsequent runs are started.
       See the debugging tutorial at https://github.com/remindmodel/remind/blob/develop/tutorials/10_DebuggingREMIND.md
Execution halted
Error in submit(cfg) : Executing prepare_and_run failed, stopping.
Execution halted
✖ | 1       0 | 1-quick [238.5s]                                                           
───────────────────────────────────────────────────────────────────────────────────────────
Failure (test_1-quick.R:3): start.R config/tests/scenario_config_quick.csv works
`status` not equal to 0.
1/1 mismatches
[1] 1 - 0 == 1
───────────────────────────────────────────────────────────────────────────────────────────
⠏ |         0 | 2-gamscompile

Reading config file config/scenario_config_AMT.csv 

No column path_gdx_refpolicycost for policy cost comparison found, using path_gdx_ref instead.

Trying to compile the selected runs...
2022-11-30 10:46:59: acquiring model lock...
acquired model lock in 0 secs.
FAIL output/gamscompile/main_SSP2EU-AMT-calibrate.lst
FAIL output/gamscompile/main_testOneRegi-AMT-Base.lst
FAIL output/gamscompile/main_SSP1-AMT-Base.lst
FAIL output/gamscompile/main_SSP1-AMT-NDC.lst
FAIL output/gamscompile/main_SSP1-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SSP1-AMT-PkBudg1320.lst
FAIL output/gamscompile/main_SSP2EU-AMT-Base.lst
FAIL output/gamscompile/main_SSP2EU-AMT-NDC.lst
FAIL output/gamscompile/main_SSP2EU-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SSP2EU-AMT-PkBudg1320.lst
FAIL output/gamscompile/main_SSP5-AMT-Base.lst
FAIL output/gamscompile/main_SSP5-AMT-NDC.lst
FAIL output/gamscompile/main_SSP5-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SSP5-AMT-PkBudg1320.lst
FAIL output/gamscompile/main_SDP-AMT-Base.lst
FAIL output/gamscompile/main_SDP-AMT-NDC.lst
FAIL output/gamscompile/main_SDP-AMT-PkBudg670.lst
FAIL output/gamscompile/main_SDP-AMT-PkBudg1320.lst

2022-11-30 10:51:20: unlocked model.

Finished: 18 runs started. 0 runs are waiting. 
To investigate potential FAILs, run: less -j 4 --pattern='^\*\*\*\*' filename.lst
Error: 18 runs were not started successfully.
Execution halted
✖ | 1       0 | 2-gamscompile [268.2s]                                                     
───────────────────────────────────────────────────────────────────────────────────────────
Failure (test_2-gamscompile.R:3): start.R --gamscompile config/scenario_config_AMT.csv works
`status` not equal to 0.
1/1 mismatches
[1] 1 - 0 == 1
───────────────────────────────────────────────────────────────────────────────────────────
✔ |     1   0 | 3-default                                                                  
───────────────────────────────────────────────────────────────────────────────────────────
Skip (test_3-default.R:2): start.R works
Reason: Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes).
───────────────────────────────────────────────────────────────────────────────────────────

══ Results ════════════════════════════════════════════════════════════════════════════════
Duration: 578.9 s

── Skipped tests  ─────────────────────────────────────────────────────────────────────────
• Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 2 | WARN 0 | SKIP 1 | PASS 1 ]
Error: Test failures
Execution halted
make: *** [Makefile:32: test] Error 1
```

Good luck finding out which tests failed and why from this wall of text.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
… continued, because wall of text.

I would suggest a couple of improvements:
- capture/suppress the output of most test, and only show it when they fail.
- stop if a test failed, do not run the others
- rearrange the tests by runtime and likelihood to fail
  - add a simple `gamscompile` on the default configuration: takes 10 seconds and if it fails, all else is moot
  - then do the entire suite of `gamscompile`, also pretty fast
  - then `codeCheck()`
  - and only then `testOneRegi`
  
That should give faster turnover.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
I have code snippets for some of this functionality, lets coordinate.

orichters commented on Makefile:
```diff
@@ -1,4 +1,4 @@
-.PHONY: help docs update-renv update-all-renv check check-fix
+.PHONY: help docs update-renv update-all-renv check check-fix test test-slow
```
please change to `test-full` as well here :)

orichters commented on tests/testthat/helper_skip_if_fast.R:
```diff
@@ -0,0 +1,8 @@
+skip_if_fast <- function() {
+  if (identical(Sys.getenv("TESTTHAT_RUN_SLOW"), "")) {
+    skip("Not run in default tests, use `make test-slow` to run (takes significantly longer than 10 minutes).")
```
`test-full` here as well.

mikapfl commented:
> I have code snippets for some of this functionality, lets coordinate.

Sounds good, I'll be busy with the R package newcomer tutorial today and tomorrow morning, will work on this PR not before tomorrow afternoon.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> > rearrange the tests by runtime and likelihood to fail
> 
> Unfortunately, a `gamscompile` only works if input data is available, so if you change the input data and run `make test` (or if you clone remind fresh and run `make test`), it wouldn't work. That's why I put all the gamscompile stuff after the `quick` stuff because the quick stuff makes sure that input data is downloaded and available.

But that should be remedied by having `start.R --gamscompile` download and distribute input data, too.  (Honestly, I thought it did that.  Bit pointless without it.)

mikapfl commented:
> But that should be remedied by having `start.R --gamscompile` download and distribute input data, too. (Honestly, I thought it did that. Bit pointless without it.)

That would be best, but is not trivial because downloading and distributing input data happens in `prepare_and_start.R`, so on cluster work nodes, typically, i.e. after the slurm submission. That is on purpose, to not do so much I/O heavy work on the login nodes. But we want gamscompile to happen on the login nodes for fast feedback. I think it would be the correct solution to refactor `prepare_and_start.R` such that the work that needs to be done for a gamscompile run is available in a separate function which can be called directly from `start.R`. But I think this is out of scope for this PR.

mikapfl commented:
I have implemented (with code parts from Michaja) that output is only printed if a test fails, and all subsequent tests are then skipped.

Additionally, I have resurrected `action=c` to do a compilation-only run which does all the data downloading and preparation if necessary and added a test `0-compile` which uses it. Unfortunately, performance is not great, it takes 55 seconds on my machine if nothing needs to be downloaded and compilation succeeds. On the other hand, it is not faster than a `quick` run until it finds the first error (both take about 40 seconds), so probably not worth it?

Anyway, output with a misplaced semi-colon in `main.gms`:

```
$ make test
Tests take about 10 minutes to run, please be patient
✔ | F W S  OK | Context
⠏ |         0 | 0-compile                                                                                                                          Command output:The following updates are available:
- mip : 0.139.1 -> 0.140.0
- mrwater : 1.4.22 -> 1.4.23
- remind2 : 1.103.21 -> 1.104.0
Consider updating with `Rscript scripts/utils/updateRenv.R`.

Reading config file config/tests/scenario_config_compile.csv 

testOneRegi
   Configuring cfg for testOneRegi
   Run can be started using 0 specified gdx file(s).
   Writing cfg to file testOneRegi.RData
    Results folder already exists, deleting and re-creating it: output/testOneRegi

The following updates are available:
- mip : 0.139.1 -> 0.140.0
- mrwater : 1.4.22 -> 1.4.23
- remind2 : 1.103.21 -> 1.104.0
Consider updating with `Rscript scripts/utils/updateRenv.R`.
   Initializing renv, see output/testOneRegi/log_renv.txt
   Writing cfg to file output/testOneRegi/config.Rdata 
   Executing prepare_and_run.R for output/testOneRegi 
Loading required package: gdx
Loading required package: gdxrrw
Loading required package: magclass
Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo 
Loading required package: mrremind
Loading required package: madrat
Loading required package: mrcommons
Loading required package: mrdrivers
Loading required package: mstools
Loading required package: edgeTransport
Loading required package: data.table

Attaching package: ‘data.table’

The following objects are masked from ‘package:dplyr’:

    between, first, last

Loading required package: mrvalidation
Loading required package: mrmagpie
Loading required package: mrland
Loading required package: mrwater

===== git info =====
Latest commit: 23951581 2022-12-05 15:01:28 +0100 Mika Pflüger 
Changes since then: On branch test
Your branch is ahead of 'origin/test' by 24 commits.
  (use "git push" to publish your local commits)

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ../../Makefile
	modified:   ../../main.gms
	modified:   ../../scripts/start/prepare_and_run.R
	modified:   ../../start.R
	modified:   ../../tests/testthat/helper_local_system2.R
	modified:   ../../tests/testthat/helper_print_if_failed.R
	modified:   ../../tests/testthat/helper_skip_failed.R
	modified:   ../../tests/testthat/test_1-quick.R

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	../../config/tests/scenario_config_compile.csv
	../../tests/testthat/test_0-compile.R
	../../todo

no changes added to commit (use "git add" and/or "git commit -a")

==== installed package versions =====
# A tibble: 27 × 2
   Package       Version 
   <chr>         <chr>   
 1 data.table    1.14.6  
 2 devtools      2.4.5   
 3 dplyr         1.0.10  
 4 edgeTransport 0.22.5  
 5 flexdashboard 0.6.0   
 6 gdx           1.51.5  
 7 gdxdt         0.1.0   
 8 gdxrrw        1.0.10  
 9 ggplot2       3.4.0   
10 gtools        3.9.4   
11 lucode2       0.39.0  
12 luplot        3.54.2  
13 luscale       2.26.5  
14 magclass      6.6.0   
15 magpie4       1.166.0 
16 mip           0.139.1 
17 mrremind      0.148.2 
18 mrvalidation  2.45.5  
19 optparse      1.7.3   
20 plotly        4.10.1  
21 remind2       1.103.21
22 rlang         1.0.6   
23 rmndt         0.5.4   
24 tidyverse     1.3.2   
25 methods       4.2.2   
26 parallel      4.2.2   
27 tools         4.2.2   

==========
Copying MAGICC files from /p/projects/rd3mod/magicc/ to results folder
2022-12-05 16:22:14: acquiring model lock...
acquired model lock in 0 secs.
No input data downloaded and distributed. To enable that, delete input/source_files.log or set cfg$force_download to TRUE.

Creating full.gms
Error in prepare() : Compiling main_testOneRegi.gms failed, stopping.
Use `less -j 4 --pattern='^\*\*\*\*' output/testOneRegi/main.lst` to investigate compilation errors.
2022-12-05 16:22:30: unlocked model.
Execution halted
Error in submit(cfg) : Executing prepare_and_run failed, stopping.
Execution halted
✖ | 1       0 | 0-compile [43.6s]                                                                                                                  
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Failure (test_0-compile.R:5): remind compiles
`status` not equal to 0.
1/1 mismatches
[1] 1 - 0 == 1
Backtrace:
 1. expect_success_status(output)
      at test_0-compile.R:5:2
 2. testthat::expect_equal(status, 0)
      at tests/testthat/helper_skip_failed.R:8:4
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
✔ |     1   0 | 1-quick                                                                                                                            
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Skip (test_1-quick.R:2): start.R config/tests/scenario_config_quick.csv works
Reason: A previous test failed.
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
✔ |     1   0 | 2-gamscompile                                                                                                                      
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Skip (test_2-gamscompile.R:2): start.R --gamscompile config/scenario_config_AMT.csv works
Reason: A previous test failed.
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
✔ |     1   0 | 3-default                                                                                                                          
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Skip (test_3-default.R:2): start.R works
Reason: Not run in default tests, use `make test-full` to run (takes significantly longer than 10 minutes).
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
✔ |     1   0 | 99-codeCheck                                                                                                                       
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Skip (test_99-codeCheck.R:4): GAMS code follows the coding etiquette
Reason: A previous test failed.
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

══ Results ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 43.7 s

── Skipped tests  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
• A previous test failed. (3)
• Not run in default tests, use `make test-full` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 1 | WARN 0 | SKIP 4 | PASS 0 ]
Error: Test failures
Execution halted
make: *** [Makefile:39: test] Error 1
```

Still quite a lot of text about skipping tests, but I think it is already a lot more useful showing only the failing test.

And a successful run is pretty short now:

```
$ make test
Tests take about 10 minutes to run, please be patient
✔ | F W S  OK | Context
✔ |         1 | 0-compile [55.4s]                                                                                                                  
✔ |         1 | 1-quick [235.1s]                                                                                                                   
✔ |         1 | 2-gamscompile [236.6s]                                                                                                             
✔ |     1   0 | 3-default                                                                                                                          
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Skip (test_3-default.R:2): start.R works
Reason: Not run in default tests, use `make test-full` to run (takes significantly longer than 10 minutes).
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
✔ |         1 | 99-codeCheck [71.9s]                                                                                                               

══ Results ════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
Duration: 599.0 s

── Skipped tests  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
• Not run in default tests, use `make test-full` to run (takes significantly longer than 10 minutes). (1)

[ FAIL 0 | WARN 0 | SKIP 1 | PASS 4 ]
```

orichters commented on tests/testthat/helper_local_system2.R:
```diff
@@ -0,0 +1,19 @@
+    # Call via system2 in remind main folder in conditions as if we were not running
+    # in tests.
+    env <- paste0(env, "unset R_PROFILE_USER;unset TESTTHAT;")
+    withr::with_dir("../..", {
```
may be an idea to move `./tests` to `./scripts/tests`…

orichters commented on config/tests/scenario_config_quick.csv:
```diff
@@ -0,0 +1,2 @@
+title;start;cm_nash_mode;cm_iteration_max;optimization;results_folder;output;cm_quick_mode;force_replace;slurmConfig
```
I think, setting `output` to `NA` does not do what is intended. At least the `quick` run shows
```
Starting output generation for output/testOneRegi


Finished output generation for output/testOneRegi!
```
instead of [this line](https://github.com/remindmodel/remind/blob/06e35960adf1d6fc6a3bab0264a0517d317c12fe/output.R#L214) catching this with `is.na(output)`.

[this line](https://github.com/remindmodel/remind/blob/develop/output.R#L214)

mikapfl replied:
It is caught instead by [this line](https://github.com/remindmodel/remind/blob/06e35960adf1d6fc6a3bab0264a0517d317c12fe/output.R#L223) because `NA.R` does not exist. Maybe good enough?

orichters replied:
Personally, I would prefer the following:
```
if (all(is.na(output) | output == "NA")) {
  message("\nNo output generation, as output was set to NA, as for example for --testOneRegi or --quick.")
```
I was very confused when I started that testOneRegi did not produce the mif file as output and didn't understand why…

orichters commented on Makefile:
```diff
@@ -32,3 +32,13 @@ check-fix:      ## Check if the GAMS code follows the coding etiquette
+
+test:           ## Test if the model compiles and runs without running a full
+                ## scenario. Tests take about 10 minutes to run.
+	$(info Tests take about 10 minutes to run, please be patient)
```
For me, was 16 minutes actually:
```
✔ | F W S  OK | Context
✔ |         1 | 0-compile [183.7s]
✔ |         1 | 1-quick [444.6s]
✔ |         1 | 2-gamscompile [363.1s]
✔ |     1   0 | 3-default
```

mikapfl replied:
On the login nodes? Because 180 s for a compilation is pretty slow, laptops should be faster. If it is on the login nodes, then there is slurm queuing and it is hard to predict how long it will take.

orichters replied:
Yes, on the login nodes…

orichters reviewed (approved):
great work, I would also merge, added some minor comments and observations, do with them whatever you like.

mikapfl commented on tests/testthat/helper_skip_if_fast.R:
```diff
@@ -0,0 +1,8 @@
+skip_if_fast <- function() {
```
yeah, there is some tension between adhering to our code standard and adhering to the code standard of testthat, which uses snake_case instead of camelCase.

mikapfl commented on tutorials/07_Advanced_ChangeCode.md:
```diff
@@ -67,30 +67,16 @@ It creates all additional gams files for your new realization "new" of the exist
 
-Before submitting the code changes to the REMIND repository, it is recommended to run `make check-fix` on the command line (or the equivalent but harder to remember `gms::codeCheck(strict = TRUE, interactive = TRUE)` in `R`). It performs some checks on the code. In particular, it may ask you some questions about the `not_used.txt` files in each realization that contain the parameter names that ar ...
-These tests are also performed on github once you submit a pull request, but it is recommended to run the tests yourself regularly during development so you don't have to fix everything at once in the end.
+The automated tests do not contain a full run of the model by default. If you do larger or more complex changes which could affect the model output, you should also do runs to compare the output of the model with and without your changes.
```
Hm, `make test-full` only tests a default run, but if people change something in some non-default realization, it won't be tested by test-full. I'll elaborate a bit more to make clear that people can use `make test-full` for a better default test, but still should make dedicated test runs for their changes.

LaviniaBaumstark reviewed (approved):
thanks! I like the new short output and helpful info on failing or skipped tests. In future we need to ensure that the scenario_config_* files used for the tests are up to date/consistent to the scenario_config- files we use for the AMT and validation. Maybe we should use one default scenario_config file and built on this one the AMT and testings.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> thanks! I like the new short output and helpful info on failing or skipped tests. In future we need to ensure that the scenario_config_* files used for the tests are up to date/consistent to the scenario_config- files we use for the AMT and validation. Maybe we should use one default scenario_config file and built on this one the AMT and testings.

Please file under #1119.
